File: scala/06-io-operations/FileIo.scala
import java.io.{FileWriter, PrintWriter}
import scala.io.Source
import scala.util.{Failure, Success, Try, Using}

object FileIo {
  def main(args: Array[String]): Unit = {
    // Write a file; Using closes the writer even if the body throws
    Using.resource(new PrintWriter("journal.txt")) { out =>
      out.println("day 1: learned println")
      out.println("day 2: learned val and var")
    }

    // Append by wrapping a FileWriter opened in append mode
    Using.resource(new PrintWriter(new FileWriter("journal.txt", true))) { out =>
      out.println("day 3: learned Using")
    }

    // Read the whole file into one String
    val contents = Using.resource(Source.fromFile("journal.txt"))(_.mkString)
    println(s"--- journal.txt (${contents.length} characters) ---")

    // Read line by line - Source is an iterator, so collection methods just work
    Using.resource(Source.fromFile("journal.txt")) { source =>
      for ((line, i) <- source.getLines().zipWithIndex)
        println(s"${i + 1}: $line")
    }

    // A missing file becomes a Failure value instead of an unhandled exception
    Try(Using.resource(Source.fromFile("missing.txt"))(_.mkString)) match {
      case Success(_) => println("missing.txt exists after all")
      case Failure(e) => println(s"could not read missing.txt: ${e.getMessage}")
    }

    new java.io.File("journal.txt").delete()
  }
}

File: rust/06-io-operations/file_io.rs
use std::fs::{self, File};
use std::io::{BufRead, BufReader, Write};

fn main() -> std::io::Result<()> {
    // Write an entire string to a file in one call (creates or truncates)
    fs::write(
        "journal.txt",
        "day 1: learned println!\nday 2: learned ownership\n",
    )?;

    // Append through an explicit handle opened with OpenOptions
    let mut file = fs::OpenOptions::new().append(true).open("journal.txt")?;
    writeln!(file, "day 3: learned the ? operator")?;

    // Read the whole file into a String at once
    let contents = fs::read_to_string("journal.txt")?;
    println!("--- journal.txt ({} bytes) ---", contents.len());

    // Read line by line with a buffered reader - each line is a Result
    let reader = BufReader::new(File::open("journal.txt")?);
    for (i, line) in reader.lines().enumerate() {
        println!("{}: {}", i + 1, line?);
    }

    // Handle a missing file gracefully instead of propagating with ?
    match fs::read_to_string("missing.txt") {
        Ok(_) => println!("missing.txt exists after all"),
        Err(e) => println!("could not read missing.txt: {}", e),
    }

    fs::remove_file("journal.txt")?;
    Ok(())
}
